### PR TITLE
paste truncated output in non pravite chats

### LINKED ISSUE #86: allow group admin set truncate policy for group, also individual could set their prefer

### PR DESCRIPTION
**Pull request by mokurin000**
fix #86, naive implementation

### CODE REVIEW

File: src/eval/execute.rs [hunk 1]
```diff
@@ -178,6 +186,20 @@ fn generate_result_from_response(resp: Response, channel: Channel, is_private: b
...
 
+async fn paste_to_pb(client: &Client, code: &str) -> reqwest::Result<String> {
+    let resp = client
+        .post("https://paste.mozilla.org/api/")
+        .form(&[
+            ("lexer", "rust"),
+            ("expires", "2073600"),
+            ("content", code),
+        ])
+        .header("content-length", 0)
```
Comment (pan93412) [inline src/eval/execute.rs:197]: I don't think this length is necessary. If this is required, I recommended you sending:

```rust
.header("content-length", code.as_bytes().len())
```

File: src/eval/execute.rs [hunk 2]
```diff
@@ -178,6 +186,20 @@ fn generate_result_from_response(resp: Response, channel: Channel, is_private: b
...
+        .send().await?
+        .text().await?;
+    Ok(resp.trim_matches('"').to_string())
```
Comment (pan93412) [inline src/eval/execute.rs:200]: I don't think it is reliable. How about pasting it as JSON?

image

File: src/eval/execute.rs [hunk 3]
```diff
@@ -178,6 +186,20 @@ fn generate_result_from_response(resp: Response, channel: Channel, is_private: b
     }
 }
 
+async fn paste_to_pb(client: &Client, code: &str) -> reqwest::Result<String> {
+    let resp = client
+        .post("https://paste.mozilla.org/api/")
+        .form(&[
+            ("lexer", "rust"),
+            ("expires", "2073600"),
```
Comment (pan93412) [inline src/eval/execute.rs:194]: By default it is `2592000` (a month). I don't think you need to modify this value manually.

File: src/eval/execute.rs [hunk 4]
```diff
@@ -112,15 +112,23 @@ fn generate_code_to_send(code: &str, bare: bool) -> String {
...
     if resp.success {
         let output = resp.stdout.trim();
-        let output = if is_private {
+        let output: Cow<'_, str> = if is_private {
             output.into()
         } else {
             const MAX_LINES: usize = 3;
             const MAX_TOTAL_COLUMNS: usize = MAX_LINES * 72;
-            utils::truncate_output(output, MAX_LINES, MAX_TOTAL_COLUMNS)
+            let mut truncated_output =
```
Comment (pan93412) [inline src/eval/execute.rs:125]: Wrap it:

```rust 
let truncated_output = {
    let mut o = utils::truncate_output(output, MAX_LINES, MAX_TOTAL_COLUMNS).into_owned();
    if let Some(pb_url) = paste_to_pb(client, output).await.ok() {
        o += "\n";
        o += &pb_url;
    }
}.into();
```

File: src/eval/execute.rs [hunk 5]
```diff
@@ -112,15 +112,23 @@ fn generate_code_to_send(code: &str, bare: bool) -> String {
...
+                utils::truncate_output(output, MAX_LINES, MAX_TOTAL_COLUMNS).into_owned();
+            if let Some(pb_url) = paste_to_pb(client, output).await.ok() {
+                truncated_output += "\n";
+                truncated_output += &pb_url;
+            }
```
Comment (pan93412) [inline src/eval/execute.rs:130]: Add a else logic, so users can understand what happened:

```rs
truncated_output += "Unable to reserve the full output log";
```

**Review by pan93412 — CHANGES_REQUESTED** [review]

File: src/eval/execute.rs (hunk 1 above)
Comment (poly000) [reply src/eval/execute.rs:197]: oh. thx
it's required but not in use
I saw curl sends length of the entire request context, thus I was really confused

File: src/eval/execute.rs (hunk 2 above)
Comment (poly000) [reply src/eval/execute.rs:200]: mozilla.paste.org gives an `"..url.."`
may I should check their documents

Comment (pan93412) [reply src/eval/execute.rs:200]: here: https://docs.dpaste.org/api/#form-parameters

By the way, `"..url.."` is also a valid JSON; therefore, do `.json::<String>()` can simplify a lot of works.

File: src/eval/execute.rs (hunk 1 above)
Comment (poly000) [reply src/eval/execute.rs]: Seems it's not a required field with correct request form

File: src/eval/execute.rs [hunk 6]
```diff
@@ -112,15 +112,39 @@ fn generate_code_to_send(code: &str, bare: bool) -> String {
...
+                    Ok(pb_url) => {
+                        let pb_url = pb_url.trim_end_matches('\n');
+                        let _ = write!(&mut o, "\n{pb_url}");
+                    }
+                    Err(e) => {
+                        let _ = write!(&mut o, "\nUnable to reserve the full output log:\n{e}");
+                    }
+                }
+                o
+            };
```
Comment (pan93412) [inline src/eval/execute.rs:145]: Better to write a unit test for this part.

File: src/eval/execute.rs [hunk 7]
```diff
@@ -178,6 +202,20 @@ fn generate_result_from_response(resp: Response, channel: Channel, is_private: b
...
+/// result will be an url with an additional '\n'
+async fn paste_to_pb(client: &Client, code: &str) -> reqwest::Result<String> {
+    const DPASTE_API_URL: &str = "https://paste.mozilla.org/api/";
+
+    let resp = client
+        .post(DPASTE_API_URL)
+        .form(&[("lexer", "rust"), ("content", code), ("format", "url")])
+        .send()
+        .await?
+        .text()
```
Comment (pan93412) [inline src/eval/execute.rs:214]: Have you resolved the API issue?

**Review by pan93412 — CHANGES_REQUESTED** [review]

File: src/eval/execute.rs (hunk 7 above)
Comment (poly000) [reply src/eval/execute.rs:214]: yeah
Content-Length is really useless
I got the missing Content Length error with an ill formed request

Comment (pan93412) [reply src/eval/execute.rs:214]: Nah, I'm talking about the response format. Can it return the paste URL without the quotes now?

Comment (poly000) [reply src/eval/execute.rs:214]: indeed
with `format` set as `url` it will simply return ".../..\n"

File: src/eval/execute.rs [hunk 8]
```diff
@@ -271,4 +310,20 @@ mod tests {
...
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(async move {
+                let example_code = "fn main(){}";
+                assert!(paste_to_pb(&client, &example_code)
+                    .await
+                    .is_ok_and(|resp| resp.starts_with(DPASTE_BASE_URL)))
+            })
+    }
```
Comment (pan93412) [inline src/eval/execute.rs:328]: Actually tokio provides [`tokio::test`](https://docs.rs/tokio/latest/tokio/attr.test.html) for testing async code. You don't need to write the initialization code manually.

**Review by pan93412 — CHANGES_REQUESTED** [review]

**Review by pan93412 — APPROVED** [review]

**Pull request closed without merging**